File: app/src/main/java/fhu/bughousechess/pieces/Bishop.java
package fhu.bughousechess.pieces;

import android.widget.ImageView;

import java.util.HashSet;
import java.util.Set;

import fhu.bughousechess.Move;
import fhu.bughousechess.R;

public class Bishop extends Piece
{
    public Bishop(String color)
    {
        this.color = color;
        type = "bishop";
        wasPawn = false;
        empty = false;
    }

    @Override
    public int getResID()
    {
        if (color.equals("white"))
            return R.mipmap.bishop;
        return R.mipmap.bbishop;
    }

    @Override
    public Set<Move> getMoves(Piece[][] positions, int x, int y, int boardNumber)
    {
        Set<Move> moves = new HashSet<>();
        boolean inbetween = false;
        int z = x;
        if (y > x)
        {
            z = y;
        }
        for (int i = 1; i < 8 - z; i++)
        {
            for (int j = 1; j < i; j++)
            {
                if (!positions[x + j][y + j].empty)
                {
                    inbetween = true;
                }
            }
            if (positions[x + i][y + i].empty)
            {
                if (!inbetween)
                {
                    moves.add(new Move(positions, x, y, x + i, y + i,"move"));
                }
            }
            if (positions[x + i][y + i].isOpposite(this))
            {
                if (!inbetween)
                {
                    moves.add(new Move(positions, x, y, x + i, y + i,"take"));
                }
            }
        }
        inbetween = false;
        if (x < y)
        {
            z = x;
        }
        else
        {
            z = y;
        }
        for (int i = 1; i < z + 1; i++)
        {
            for (int j = 1; j < i; j++)
            {
                if (!positions[x - j][y - j].empty)
                {
                    inbetween = true;
                }
            }
            if (positions[x - i][y - i].empty)
            {
                if (!inbetween)
                {
                    moves.add(new Move(positions, x, y, x - i, y - i,"move"));
                }
            }
            if (positions[x - i][y - i].isOpposite(this))
            {
                if (!inbetween)
                {
                    moves.add(new Move(positions, x, y, x - i, y - i,"take"));
                }
            }
        }
        inbetween = false;
        if (7 - x < y)
        {
            z = 7 - x;
        }
        else
        {
            z = y;
        }
        for (int i = 1; i < z + 1; i++)
        {
            for (int j = 1; j < i; j++)
            {
                if (!positions[x + j][y - j].empty)
                {
                    inbetween = true;
                }
            }
            if (positions[x + i][y - i].empty)
            {
                if (!inbetween)
                {
                    moves.add(new Move(positions, x, y, x + i, y - i,"move"));
                }
            }
            if (positions[x + i][y - i].isOpposite(this))
            {
                if (!inbetween)
                {
                    moves.add(new Move(positions, x, y, x + i, y - i,"take"));
                }
            }
        }
        inbetween = false;
        if (7 - y < x)
        {
            z = 7 - y;
        }
        else
        {
            z = x;
        }
        for (int i = 1; i < z + 1; i++)
        {
            for (int j = 1; j < i; j++)
            {
                if (!positions[x - j][y + j].empty)
                {
                    inbetween = true;
                }
            }
            if (positions[x - i][y + i].empty)
            {
                if (!inbetween)
                {
                    moves.add(new Move(positions, x, y, x - i, y + i,"move"));
                }
            }
            if (positions[x - i][y + i].isOpposite(this))
            {
                if (!inbetween)
                {
                    moves.add(new Move(positions, x, y, x - i, y + i,"take"));
                }
            }
        }
        return moves;
    }
}

File: app/src/main/java/fhu/bughousechess/pieces/Bishop.kt
package fhu.bughousechess.pieces

import fhu.bughousechess.Move
import java.util.HashSet

class Bishop(color: String?) : Piece() {
    init {
        this.color = color
        type = "bishop"
        wasPawn = false
        empty = false
    }

    override fun getMoves(
        positions: Array<Array<Piece>>,
        x: Int,
        y: Int,
        boardNumber: Int
    ): Set<Move?>? {
        val moves: MutableSet<Move?> = HashSet()
        var inbetween = false
        var z = x
        if (y > x) {
            z = y
        }
        for (i in 1 until 8 - z) {
            for (j in 1 until i) {
                if (!positions!![x + j]!![y + j]!!.empty) {
                    inbetween = true
                }
            }
            if (positions!![x + i]!![y + i]!!.empty) {
                if (!inbetween) {
                    moves.add(Move(positions, x, y, x + i, y + i, "move"))
                }
            }
            if (positions[x + i]!![y + i]!!.isOpposite(this)) {
                if (!inbetween) {
                    moves.add(Move(positions, x, y, x + i, y + i, "take"))
                }
            }
        }
        inbetween = false
        z = if (x < y) {
            x
        } else {
            y
        }
        for (i in 1 until z + 1) {
            for (j in 1 until i) {
                if (!positions!![x - j]!![y - j]!!.empty) {
                    inbetween = true
                }
            }
            if (positions!![x - i]!![y - i]!!.empty) {
                if (!inbetween) {
                    moves.add(Move(positions, x, y, x - i, y - i, "move"))
                }
            }
            if (positions[x - i]!![y - i]!!.isOpposite(this)) {
                if (!inbetween) {
                    moves.add(Move(positions, x, y, x - i, y - i, "take"))
                }
            }
        }
        inbetween = false
        z = if (7 - x < y) {
            7 - x
        } else {
            y
        }
        for (i in 1 until z + 1) {
            for (j in 1 until i) {
                if (!positions!![x + j]!![y - j]!!.empty) {
                    inbetween = true
                }
            }
            if (positions!![x + i]!![y - i]!!.empty) {
                if (!inbetween) {
                    moves.add(Move(positions, x, y, x + i, y - i, "move"))
                }
            }
            if (positions[x + i]!![y - i]!!.isOpposite(this)) {
                if (!inbetween) {
                    moves.add(Move(positions, x, y, x + i, y - i, "take"))
                }
            }
        }
        inbetween = false
        z = if (7 - y < x) {
            7 - y
        } else {
            x
        }
        for (i in 1 until z + 1) {
            for (j in 1 until i) {
                if (!positions!![x - j]!![y + j]!!.empty) {
                    inbetween = true
                }
            }
            if (positions!![x - i]!![y + i]!!.empty) {
                if (!inbetween) {
                    moves.add(Move(positions, x, y, x - i, y + i, "move"))
                }
            }
            if (positions[x - i]!![y + i]!!.isOpposite(this)) {
                if (!inbetween) {
                    moves.add(Move(positions, x, y, x - i, y + i, "take"))
                }
            }
        }
        return moves
    }
}